### ISSUE TITLE
`prefer-const` - Linter does not recognized variable is being reassigned in function parameter

### ISSUE THREAD

**Issue by apacheli** [denoland/deno_lint#1065]
### Lint Name

prefer-const

### Code Snippet

```ts
export let x = 0;

export const funcA = (someNumber: number = x++) => {
  return someNumber;
};

export const funcB = (someNumber: number = x = x + 1) => {
  return someNumber;
};

export const funcC = (someNumber: number = x += 1) => {
  return someNumber;
};

```

### Expected Result

Linter should not recommend `prefer-const` rule as `x` can possibly be reassigned.

### Actual Result

Linter does not think `x` is being reassigned, thus recommending that I use `prefer-const`.

### Version

1.24.2

**Comment by kitsonk** [comment]
I could have sworn this used to work, but no, it doesn't.  another reproduction:

```ts
function b(): [string] {
  return ["hello"];
}

function a() {
  let a: string | undefined;
  [a] = b();
}
```

It incorrectly says: `"a" is never reassigned Use "const" instead`.

**Comment by ah-yu** [comment]
> I could have sworn this used to work, but no, it doesn't. another reproduction:
> 
> ```ts
> function b(): [string] {
>   return ["hello"];
> }
> 
> function a() {
>   let a: string | undefined;
>   [a] = b();
> }
> ```
> 
> It incorrectly says: `"a" is never reassigned Use "const" instead`.

This is expected behavior, right? You only assign `a` once.

**Comment by kitsonk** [comment]
> > I could have sworn this used to work, but no, it doesn't. another reproduction:
> > ```ts
> > function b(): [string] {
> >   return ["hello"];
> > }
> > 
> > function a() {
> >   let a: string | undefined;
> >   [a] = b();
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It incorrectly says: `"a" is never reassigned Use "const" instead`.
> 
> This is expected behavior, right? `a` is assigned only once.

The is re-assigned once, it's initial value is `undefined` and it is assigned the value of `"hello"`.  With blocked scoped `let` and `const` you can't always do the assignment within the same block, and even if you put another block there, the linter still complains, and it should.  It shouldn't have opinions about why you would declare the variable and then assign it a value.

**Comment by ah-yu** [comment]
I see. Thank you for the explanation.

**Comment by ah-yu** [comment]
After looking at the documentation, I found that variables allow conditional initialization. I tested with your example and this lint rule doesn't complain.
```js
// conditionally initialized variable is allowed
let h;
if (trueOrFalse) {
  h = 0;
}
```

I don't know if I misunderstood you. If we declare a variable, the variable is considered to be assigned an initial value. But it's different from what the documentation says (see below).
```js
// variable that is uninitialized at first and then assigned in the same scope is NOT allowed
// because we could simply write it like `const e = 2;` instead
let e;
e = 2;
```
This means we can simplify `let a; [a] = b()` to `const [a] = b()`, so lint works as expected, right? Or do you think this lint shouldn't complain about the above?
